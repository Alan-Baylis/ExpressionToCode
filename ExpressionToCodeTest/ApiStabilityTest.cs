using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using ExpressionToCodeLib;
using Xunit;

namespace ExpressionToCodeTest
{
    public class ApiStabilityTest
    {
        [Fact]
        public void PublicApi()
        {
            var publicTypes = typeof(ExpressionToCode).Assembly.GetTypes()
                .Where(IsPublic)
                .Where(type => !type.Namespace.Contains("Unstable"))
                .OrderByDescending(type => type.IsEnum)
                .ThenByDescending(type => type.IsInterface)
                .ThenBy(type => type.FullName);

            ApprovalTest.Verify(PrettyPrintTypes(publicTypes));
        }

        [Fact]
        public void UnstableApi()
        {
            var unstableTypes = typeof(ExpressionToCode).Assembly.GetTypes()
                .Where(IsPublic)
                .Where(type => type.Namespace.Contains("Unstable"))
                .OrderByDescending(type => type.IsEnum)
                .ThenByDescending(type => type.IsInterface)
                .ThenBy(type => type.FullName);

            ApprovalTest.Verify(PrettyPrintTypes(unstableTypes));
        }

        static string PrettyPrintTypes(IEnumerable<Type> types) => string.Join("", types.Select(PrettyPrintTypeDescription));
        static string PrettyPrintTypeDescription(Type o) => PrettyPrintTypeHeader(o) + "\n" + PrettyPrintTypeContents(o);

        static string PrettyPrintTypeContents(Type type)
        {
            var methods = type.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                    .OrderBy(mi => mi.MetadataToken)
                    .Where(mi => mi.DeclaringType.Assembly != typeof(object).Assembly) //exclude noise
                ;

            var methodBlock = string.Join("", methods.Select(mi => PrettyPrintMethod(mi) + "\n"));

            var fields = type.GetFields(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                    .Where(mi => mi.DeclaringType.Assembly != typeof(object).Assembly) //exclude noise
                ;

            var fieldBlock = string.Join("", fields.Select(fi => PrettyPrintField(fi) + "\n"));

            return fieldBlock + methodBlock + "\n";
        }

        static string PrettyPrintTypeHeader(Type type)
        {
            var prefix = TypePrefix(type);

            var baseType = type.BaseType == typeof(object) ? null : type.BaseType;
            var allInterfaces = type.GetInterfaces();
            var interfaces = baseType == null ? allInterfaces : allInterfaces.Except(baseType.GetInterfaces());
            var inheritanceTypes = new[] { baseType }.OfType<Type>().Concat(interfaces);
            var suffix = !inheritanceTypes.Any() || type.IsEnum ? "" : " : " + string.Join(", ", inheritanceTypes.Select(ObjectToCode.ToCSharpFriendlyTypeName));

            var name = ObjectToCode.ToCSharpFriendlyTypeName(type);

            return prefix + " " + name + suffix;
        }

        static string TypePrefix(Type type)
        {
            if (type.IsEnum) {
                return "enum";
            } else if (type.IsValueType) {
                return "struct";
            } else if (type.IsInterface) {
                return "interface";
            } else {
                return "class";
            }
        }

        static string PrettyPrintMethod(MethodInfo mi)
        {
            var fakeTarget = mi.IsStatic ? "TYPE" : "inst";

            return "    " + ObjectToCode.ToCSharpFriendlyTypeName(mi.ReturnType) + " " + fakeTarget +
                "." + mi.Name
                + PrettyPrintGenericArguments(mi)
                + PrettyPrintParameterList(mi);
        }

        static object PrettyPrintField(FieldInfo fi) => "    "
            + (fi.IsLiteral ? "const " : (fi.IsStatic ? "static " : "") + (fi.IsInitOnly ? "readonly " : ""))
            + ObjectToCode.ToCSharpFriendlyTypeName(fi.FieldType)
            + " " + fi.Name
            + (fi.IsLiteral ? " = " + ObjectToCode.ComplexObjectToPseudoCode(fi.GetRawConstantValue()) : "");

        static string PrettyPrintParameterList(MethodInfo mi)
        {
            return "(" + string.Join(
                ", ",
                mi.GetParameters()
                    .Select(
                        pi =>
                            ObjectToCode.ToCSharpFriendlyTypeName(pi.ParameterType) + " " + pi.Name)) + ")";
        }

        static string PrettyPrintGenericArguments(MethodInfo mi)
        {
            if (!mi.IsGenericMethodDefinition) {
                return "";
            }
            return "<"
                + string.Join(", ", mi.GetGenericArguments().Select(ObjectToCode.ToCSharpFriendlyTypeName))
                + ">";
        }

        static bool IsPublic(Type type) => type.IsPublic || type.IsNestedPublic && IsPublic(type.DeclaringType);
    }
}
