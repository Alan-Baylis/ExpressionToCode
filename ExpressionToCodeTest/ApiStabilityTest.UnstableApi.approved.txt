interface IExpressionToCode
    string inst.ToCode(Expression e)

class ExpressionStringify : IExpressionToCode
    string inst.ToCode(Expression e)
    IExpressionToCode TYPE.With(bool fullTypeNames, bool explicitMethodTypeArgs)

class ExpressionToCodeExt
    string TYPE.ToCode<T, T1, T2, T3>(IExpressionToCode it, Expression<Func<T, T1, T2, T3>> e)
    string TYPE.ToCode<T, T1, T2>(IExpressionToCode it, Expression<Func<T, T1, T2>> e)
    string TYPE.ToCode<T, T1>(IExpressionToCode it, Expression<Func<T, T1>> e)
    string TYPE.ToCode<T>(IExpressionToCode it, Expression<Func<T>> e)

class ExpressionWithName
    string TYPE.ToNameOf<TResult>(Expression<Func<TResult>> expression)
    string TYPE.ToNameOf<T1, T2, TResult>(Expression<Func<T1, T2, TResult>> expression)
    string TYPE.ToNameOf<T1, TResult>(Expression<Func<T1, TResult>> expression)
    string TYPE.ToNameOf<T1, T2, T3, TResult>(Expression<Func<T1, T2, T3, TResult>> expression)
    string TYPE.ToNameOf<T>(Expression<Action<T>> expression)
    string TYPE.ToNameOf<T1, T2>(Expression<Action<T1, T2>> expression)
    string TYPE.ToNameOf<T1, T2, T3>(Expression<Action<T1, T2, T3>> expression)
    string TYPE.ToNameOf(Expression<Action> expression)
    string TYPE.ToNameOf<T>(Expression<T> expression)
    string TYPE.ToFullNameOf<T>(Expression<T> expression)

class ObjectStringifierExtensions
    string TYPE.PlainObjectToCode(IObjectStringifier it, object val)

class ObjectStringify
    static readonly IObjectStringifier Default
    static readonly IObjectStringifier WithFullTypeNames

class ExpressionWithValue
    string TYPE.ToValuedCode<TResult>(Expression<Func<TResult>> expression)

